### ISSUE TITLE
Cache memory protection / SELinux / Apple Hardened Runtime

### ISSUE THREAD

**Issue by richard-gravy** [dosbox-staging/dosbox-staging#253]
Hello,

So after running a DOS program for a little while, I got unceremoniously dumped out of Dosbox- staging with the following messages in the console:-
`DOSBox has switched to max cycles, because of the setting: cycles=auto.`
`If the game runs too fast, try a fixed cycles amount in DOSBox's options.`
`Setting execute permission on the code cache has failed!`
`Segmentation fault (core dumped)`
I also received a message that an AVC denial has occured.

The source of the problem seems to be https://github.com/dreamer/dosbox-staging/blob/f2029d71d8d97e54358dbb0610643a879468f22a/src/cpu/core_dyn_x86/cache.h#L536, which allocates cache storage from the heap and later on calls mprotect() to make it executable.
According to the gospel of SELinux, this is a "no".

It's worth noting that the Windows-specific code doesn't allocate from the heap, but uses VirtualAlloc() to allocate pages of virtual memory for this purpose.  The Linux* equivalent is mmap().

Can this be corrected, rather than rolling out the "Change SELinux to permissive mode" trope?
\
\
*Well, yeah, it's BSD but was recently blessed by POSIX.

**Comment by krcroft** [comment]
Interesting, and thanks for sharing a suggested fix for this @richard-gravy 

I guess SELinux permits memory to be made executable by mmap, but not mprotect? (virus authors just switch to mmap and carry on? maybe mmap'ed memory is more strictly handled by the kernel or SELinux?.. just wonder out loud :-)

If you replace the line you flagged with this, does it satisfy SELinux?

 `cache_code_start_ptr = static_cast<Bit8u*>(mmap(0, CACHE_TOTAL + CACHE_MAXSIZE + PAGESIZE_TEMP - 1 + PAGESIZE_TEMP, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);`

**Comment by richard-gravy** [comment]
The issue is the use of malloc() specifically, which allocates from the heap.  I suppose that as you could theoretically return addresses for separate small allocation calls referring to different parts of the same page (as was done when programs did what they could with 4MB of RAM), changing the protection of a heap page could have unintended effects.

In the modern era, it might be easier to alias malloc() to mmap().

**Comment by krcroft** [comment]
Thanks for the explanation @richard-gravy.

**Comment by dreamer** [comment]
@richard-gravy eh; yes - this is a problem. There is a patch on Vogons forum, that should address this (it's a massive patch) and seemed to work when I tested it, but:

- I don't have access to computer where I have that patch imported and rebased on top of dosbox-staging master - and it seems like I forgot to push it as a `vogons/*` branch to this repo…
- That patch is *massive* - it makes certain design decisions I don't agree with, and overall needs serious cleaning and limitation of scope before passing criteria for inclusion in dosbox-staging. Perhaps the change will need to be rewritten from grounds-up.
- I don't have enough experience in CPU emulation area yet to understand all intricacies, that might be at play for this change… it will take some time.

So, stay tuned, I will post WIP version of that patch in the evening for you to test.

**Comment by dreamer** [comment]
Ah, forgot to link the WIP patch earlier - it is ed27c060b699f604b91481c2b65c3ef5476b52e4 although it does not apply cleanly on newest master any more (but conflicts seem easy to resolve).

**Comment by dreamer** [comment]
I am not an expert in cpu emulation area, I spent some time today reviewing code in there, and have some conclusions on how to proceed with this task.

This is bigger than it looks and very delicate; the big source of the problem is duplicated implementation of dynamic recompiler: there's `core_dyn_x86` and `core_dynarec`. Both implementations use `mprotect`.

As far as I can tell, `core_dyn_x86` is the older one, at some point, DOSBox maintainers copied the code and started `core_dynarec` (which handles multiple architectures). `core_dyn_x86` is faster one (I think), enabled by default on x86 architectures, and is receiving more attention from upstream nowadays.

I started reviewing **jmarsh**'s implementation: ed27c060b699f604b91481c2b65c3ef5476b52e4 - we can't use it any more, unfortunately, because the patch has multiple code conflicts (seemingly both with SVN trunk and our master branch). Also, that patch was never properly tested outside of Linux x86_64.

Overall the implementation in there seems fine (I haven't finished reviewing it yet). **jmarsh** also made the *correct* call of removing duplicated implementation between `core_dyn_x86/cache.h` and `core_dynrec/cache.h`. However, that decision is the source of conflicts that prevent us from easily merging it right now. These two files are ~70%-75% the same, and after adjusting for name changes maybe ~80%.

But there's another problem - these changes conflict (functionally) with *another* **jmarsh** patch, which is waiting for inclusion in SVN trunk for much longer now: 26ecc855c10b7b2b3b7eaee421e9a2c5d67c8ade - this one was actually tested and seems to work well.

So, I am not 100% sure what is the best way to proceed with this, but here's my plan for now:

1. Start removing duplication in `cache.h` files, similarly like jmarsh did it - except instead of one "big" replacement, do it gradually. After this step we'll end up with **3** cache.h files (instead of 2 files right now): `src/cpu/cache.h` (the part with stable, shared parts) and `src/cpu/core_dyn_x86/cache.h` and `src/cpu/core_dynrec/cache.h` (with parts specific to those cores). Aside of code moved around, this should have no functional changes and no regressions.

2. (perhaps parallel to 1.) merge-in PPC dynarec patch
3. Implement facede class similar to `cache_pos` class
https://github.com/dosbox-staging/dosbox-staging/blob/ed27c060b699f604b91481c2b65c3ef5476b52e4/src/cpu/cache.h#L79 - I particularly do not like over-reliance on operator overloading in there.
4. Finally, with all those pieces in place (and knowledge gained), we'll be able to replace `mprotect` (once); this task in itself won't be easy, but we have one reference implementation already, and other resources as well - e.g. [link](https://www.talospace.com/2020/01/dosbox-jit-on-ppc64le-and-how-you-can.html).

*Similarly to SDL2 work - if someone has better idea on how to proceed about this, then please speak up.*

**Comment by kcgen** [comment]
@jmarsh, @qbix: DOSBox no longer runs on default installations of Debian and Fedora due to this writable/executable memory security block.  

Getting the dynamic core working on common Linux platforms is a task exclusive of project goals, dev tools, and processes. If you have comments or the approach mentioned by @dreamer  above, feel free to PM me on Vogons, which I will relay here. Thank you!

**Comment by richard-gravy** [comment]
I have re-read what I wrote a few times and I think clarification is required.

**mprotect** is not the source of the problem.  The use of **malloc**/**free** is.  The reason is that the heap is something that a C/C++ runtime provides, **not** the operating system.  When you make a malloc() call for say 300 bytes, the C runtime internally checks if it has pages available to satisfy your request.  If not, it allocates a page using mmap(), specifying "no execute" in the protection flags, does some bookkeeping and then hands you an address good for 300 bytes.  You may be able to read/write more than 300 bytes, though most runtimes place "canaries" in the free space left in pages to detect buffer overflows, which will trigger when using free() or during a debug process.

In essence, the heap abstracts away the concept of memory allocation from something the platform understands (pages of about 4KB) to individual blocks of bytes.

By using **mprotect** on heap pages, you are changing the protection of pages that the C/C++ runtime is supposed to be abstracting _away_ from your control.  It has allocated them and is using them for more than storing your bytes.  You are using them in an _unintended_ way.

In modern Linux, this behaviour can be seen by the kernel and is mostly blocked by default.  If you want to deal with pages, allocate them from the OS yourself.

**Comment by dreamer** [comment]
Small status update: until now I was hoping I'll find time to fix this issue before 0.76.0 release… but I don't want to postpone the release - we have fair number of various new features and improvements - holding the release back won't help anyone.

Few weeks ago I made significant progress towards fixing this issue - refector/cleanup in caches is now 99% complete and merged in to master branch. There is still some work to do - it's mostly tedious and very, very risky. Not, it's not as easy as replacing `memprotect` with `mmap` (the separate pointers to writeable/executable caches still need to be used in appropriate places).

Even after merging, we will need to spend a lot of time testing on various architectures…

**I'm moving this bug out of 0.76.0 release; new target release for this fix is 0.77.0.**

**Comment by kcgen** [comment]
Apple's ARM-based laptop has brought this issue to a head. Active discussion w/ @joncampbell123 and jmarsh: https://www.vogons.org/viewtopic.php?p=914518#p914518

**Comment by joncampbell123** [comment]
@kcgen Another issue brought forth by DOSBox-X is a problem with Visual Studio 2019, 64-bit builds (for Intel), and the dynamic x86 core. A regression causes it to fail to load the address of &cpu_regs and a crash. Has that issue happened yet with this project?

**Comment by joncampbell123** [comment]
@kcgen @dreamer DOSBox-X's current workaround happens to work, but it's suboptimal. It calls mprotect every time it needs to add a cache block, then mprotect again to make it executable.

I'm told that Mac OS X allows vm_remap() to make two different mappings to the same memory, one read+write and one read+execute. Someone linked a PDF where supposedly an iOS framework uses that to JIT compile JavaScript. I would like to use it, despite the absolutely abysmal documentation surrounding that function and other core Mach/Darwin kernel functions of that kind.

**Comment by kcgen** [comment]
@joncampbell123, thanks for pulling us in, and I should say you wasted no time in picking up a new Apple ARM system! All the better to get this issue tackled. 

> Has (the VS2019 register-related) issue happened yet with this project?

We extensively build 64-bit snapshots using VS2019 and test them, and a good portion of the community uses them too, but we haven't hit it.

 DOSBox Staging's CPU core code tightly parallels SVN and can be considered functionality equivalent, however @dreamer's refactored a decent chunk of the cache and related code (largely duplicated across multiple files) by collapsing it into one common area, which will be the basis to work the W^X solution.

I haven't spent much time in that code so just giving intro context.. @dreamer will have much more to discuss 👍

**Comment by joncampbell123** [comment]
@kcgen I'm told @Wengier wasn't hitting the issue either, despite compiling with VS2019 64-bit with the same codebase. Curious. One possible theory is that you and @Wengier were getting compiles where &cpu_regs referred to a memory address below 4GB, or the VS2019 typecast regression responsible didn't occur for anyone except me.

**Comment by kcgen** [comment]
> One possible theory is that you and @Wengier were getting compiles where &cpu_regs referred to a memory address below 4GB

Interesting; if this issue originates (or can be obviated) at compile-time, then perhaps a `static_assert` with your check would hit it mid-build.

(Ref: https://github.com/joncampbell123/dosbox-x/commit/15d5114d62b277f48a178edff39886b65bdbb3c9)

**Comment by grapeli** [comment]
> Future plans discussed for DOSBox-X are that, if a Linux distribution is enforcing W^X, to try creating a memfd and mapping that twice the same manner (which would require a Linux kernel as recent as 4.1.20, I think). memfds are theoretically not affected by the mount options of /dev/shm in case any Linux distribution has mounted it noexec.

It is certainly not kernel 4.1.20, but 3.17 and newer.
https://akkadia.org/drepper/selinux-mem.html
https://kernelnewbies.org/Linux_3.17#A.27File_sealing.27_eases_handling_of_shared_memory

An open system would force something on the user (he will always have a choice). Maybe on ChromeOS or Android.

**Comment by dreamer** [comment]
One note to remember: to trigger the problem user needs to actually test a DOS program that will trigger DOSBox to switch to a dynamic core. So testing shell, simple text programs and even simple games is not enough - user needs to actually run "heavier" DOS game or program. *This might be the major reason why false-negative test results might be reported.*

For example: Norton Commander would not trigger the switch to dynamic core, thus the user would not even notice there's a bug. But DOS Navigator should trigger the memory protection immediately.

@joncampbell123 

I haven't tested **jmarsh**'s patch on Windows at all. As for implementation - he's approach is based on using:

1. (POSIX): use `shm_open` (or fall back to `mkstemp` if it's unavailable at compile time), then two `mmap` calls to obtain W pointer and X pointer
2. (Windows): use `CreateFileMapping`, then two `MapViewOfFile` calls to obtain W pointer and X pointer
3. (cross-platform): Store the diff between W and X pointers (for converting W pointer to X pointer as needed) and then annotate/redirect all uses of cache pointers to use either W or X (without going into details how it is implemented...)

Steps (1) and (2) are equivalent (which is nice) - I remember reading that it is a recommended way for implementing safe JIT mechanisms (I think I read a blog posts from people working on JavaScript JIT in Firefox? Not sure, can't find it right now).

Steps (1) and (2) are not problematic - as far as I can tell - it works exactly as it should. Step (3) it the problematic one - storing the diff is a clever and convenient solution, but I'm not sure if it's a good one; annotating code is the big hurdle to overcome in my opinion - that's what I am struggling with (**jmarsh** uses C-style casting and overloading of casting operator to achieve this).

**Comment by dreamer** [comment]
Added macOS label, as this is a cross-platform issue now. Apple's documentation: https://developer.apple.com/documentation/apple_silicon/porting_just-in-time_compilers_to_apple_silicon?language=objc

**Comment by kklobe** [comment]
Quick note that AS uses a 16k pagesize. Right now I'm just doing a quick hack in configure.ac but that probably won't work well for a universal build.

**Comment by joncampbell123** [comment]
You may also look at the changes DOSBox-X has made to dynamic core to support Mac OS X and SELinux.

It supports multiple methods of memory mapping and can support the dual read/write + read/execute mapping scheme.

Because of how dynamic core works, some changes to x86 and x86_64 is also needed to direct branches, calls, and jumps to the executable region instead of the read/write region.

**Comment by joncampbell123** [comment]
As for Windows... I'm not entirely sure you need to do dual mapping. DOSBox-X dynamic core still worked perfectly fine even when I went into the System settings and enabled Data Execution Prevention for all programs. Also my past experience with file memory mapping in Windows is that Microsoft's memory manager was extraordinarily stupid picky about memory mapping (though of course this was back in the Windows XP days) so I'm not sure you need to worry about that in Windows yet.

**Comment by dreamer** [comment]
On Windows we're experiencing crashes in CPU dynamic core, but only for 64-bit, only release builds (that's why we took them down for now) and only when building with Visual Studio **16.8.2** (it was released 2 weeks ago - older VS did not generate crashing builds): https://github.com/dosbox-staging/dosbox-staging/issues/746#issuecomment-738686492

It sure looks like a memory protection issue (maybe Microsoft silently turned some switch on recently?), but I'm not 100% convinced it is yet (might be e.g. a bug in FPU emulator, as 64-bit uses different implementation than 32-bit on Windows), or just some regression in MSVC (yeah, ridiculous for me to suggest it, I know).

**Comment by joncampbell123** [comment]
@dreamer I know exactly why dynamic core crashes with recent VS2019 builds. There is code in x86 and x86_64 dynamic core to check whether the address of &cpu_regs is above or below 4GB (whether the pointer can fit in 32 bits or not). It relies on typecasting to Bitu and then Bit32u and comparing against itself. For some reason, VS2019 optimizes it out completely into taking the "it's small enough to fit in 32 bits" branch even if clearly not. That affects whether dynamic core generates a "mov ebp,&cpu_regs" or a "mov rbp,&cpu_regs".

https://github.com/joncampbell123/dosbox-x/commit/15d5114d62b277f48a178edff39886b65bdbb3c9

**Comment by dreamer** [comment]
@joncampbell123 thank you! I'll try to port your fix :)

**Comment by joncampbell123** [comment]
It's almost as subtle as what Clang/LLVM does to the funny "subtract and branch if carry" noise generator stuff in the PC-98 FM board code. Why, take the 32-bit signed ints, sign extend to 64-bit, subtract, and then see if that carries. No, of course that's not what the code intended, why does the noise channel not work? :)

**Comment by kklobe** [comment]
resolved by https://github.com/dosbox-staging/dosbox-staging/pull/1031

**Comment by excitoon** [comment]
```
core = simple
```

**Comment by excitoon** [comment]
Reproduced by running Borland C++ 3.0
